sdk/core/core/src/az_curl_adapter.c

```diff
@@ -10,27 +10,218 @@
 
 typedef struct {
   struct curl_slist * p_list;
-} az_headers_data;
+} az_curl_headers_list;
 
-AZ_CALLBACK_FUNC(az_headers_to_curl, az_headers_data *, az_pair_visitor)
+/**
+ * @brief writes a header key and value to a buffer as a 0-terminated string and using a separator
+ * span in between. Returns error as soon as any of the write operations fails
+ *
+ * @param writable_buffer
+ * @param p_header
+ * @param separator
+ * @return az_result
+ */
+az_result az_write_to_buffer(
+    az_span const writable_buffer,
+    az_pair * const p_header,
+    az_const_span separator) {
+  az_span_builder writer = az_span_builder_create(writable_buffer);
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, p_header->key));
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, separator));
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, p_header->value));
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, AZ_STR("\0")));
+  return AZ_OK;
+}
+
+/**
+ * @brief allocate a buffer for a header. Then reads the az_pair header and writes a buffer. Then
+ * uses that buffer to set curl header. Header is set only if write operations were OK. Buffer is
+ * free after setting curl header.
+ *
+ * @param p_header
+ * @param p_headers
+ * @param separator
+ * @return az_result
+ */
+az_result az_add_header_to_curl_list(
+    az_pair * const p_header,
+    az_curl_headers_list * const p_headers,
+    az_const_span separator) {
+  // allocate a buffet for header
+  int16_t buffer_size = p_header->key.size + separator.size + p_header->value.size + 1;
+  uint8_t * const p_writable_buffer = (uint8_t * const)malloc(buffer_size);
```

Comment:
We do malloc in several places for CURL, I would suggest to have a function that returns `az_span` or `az_span_builder`. For example `az_result az_span_new(size_t size, az_span *out)`.

Reply:
Should we use `alloc`/`free` suffixes or `new`/`delete` ?

Reply:
I vote for alloc

Reply:
suffixes even if this is part of an adapter that we now in advance that it's allocating?  seems like on those cases we don't really care if the adapter is allocating or not

Reply:
The problem is not suffixes but multiple malloc/free in different places. These places are very hard to prof that these is no leak, dangling pointer, security issue etc.